rust/headless-client/src/windows.rs

```diff
@@ -0,0 +1,487 @@
+        WinSec::InitializeSecurityDescriptor(
+            psd,
+            windows::Win32::System::SystemServices::SECURITY_DESCRIPTOR_REVISION,
+        )
+        .context("InitializeSecurityDescriptor failed")?;
+        WinSec::SetSecurityDescriptorDacl(psd, true, None, false)
+            .context("SetSecurityDescriptorDacl failed")?;
+    }
+
+    let mut sa = WinSec::SECURITY_ATTRIBUTES {
+        nLength: 0,
+        lpSecurityDescriptor: psd.0,
+        bInheritHandle: false.into(),
+    };
+    sa.nLength = std::mem::size_of_val(&sa)
+        .try_into()
+        .context("Size of SECURITY_ATTRIBUTES struct is not right")?;
+
+    let sa_ptr = &mut sa as *mut _ as *mut c_void;
+    // SAFETY: Unsafe needed to call Win32 API. We only pass pointers to local vars, and Win32 shouldn't store them, so there shouldn't be any threading of lifetime problems.
+    let server = unsafe { server_options.create_with_security_attributes_raw(pipe_path(), sa_ptr) }
+        .context("Failed to listen on named pipe")?;
+    Ok(server)
+}
+
+/// Named pipe for IPC between GUI client and IPC service
+pub fn pipe_path() -> String {
+    named_pipe_path(&format!("{BUNDLE_ID}.ipc_service"))
+}
+
+enum IpcEvent {
+    /// A message that the client sent us
+    Client(IpcClientMsg),
+    /// A message that connlib wants to send
+    Connlib(IpcServerMsg),
+    /// The IPC client disconnected
+    IpcDisconnect,
+}
+
+#[derive(Clone)]
+struct CallbackHandlerIpc {
+    cb_tx: mpsc::Sender<IpcServerMsg>,
+}
+
+impl Callbacks for CallbackHandlerIpc {
+    fn on_disconnect(&self, _error: &connlib_client_shared::Error) {
+        self.cb_tx
+            .try_send(IpcServerMsg::OnDisconnect)
+            .expect("should be able to send OnDisconnect");
+    }
+
+    fn on_set_interface_config(
+        &self,
+        ipv4: Ipv4Addr,
+        ipv6: Ipv6Addr,
+        dns: Vec<IpAddr>,
+    ) -> Option<i32> {
+        tracing::info!("TunnelReady");
+        self.cb_tx
+            .try_send(IpcServerMsg::OnSetInterfaceConfig { ipv4, ipv6, dns })
+            .expect("Should be able to send TunnelReady");
+        None
+    }
+
+    fn on_update_resources(&self, resources: Vec<callbacks::ResourceDescription>) {
+        tracing::info!(len = resources.len(), "New resource list");
+        self.cb_tx
+            .try_send(IpcServerMsg::OnUpdateResources(resources))
+            .expect("Should be able to send OnUpdateResources");
+    }
+}
+
+async fn handle_ipc_client(server: named_pipe::NamedPipeServer) -> Result<()> {
+    let framed = Framed::new(server, LengthDelimitedCodec::new());
+    let mut framed = pin!(framed);
+    let (cb_tx, mut cb_rx) = mpsc::channel(100);
+
+    let mut connlib = None;
+    let callback_handler = CallbackHandlerIpc { cb_tx };
+    loop {
+        let ev = poll_fn(|cx| {
+            match cb_rx.poll_recv(cx) {
+                Poll::Ready(Some(msg)) => return Poll::Ready(Ok(IpcEvent::Connlib(msg))),
```

Comment:
Same as for the other one, I think you could split this into two tasks where you are directly forwarding from the receiver to the `FramedWrite`.

Reply:
I must have done something like that, all the `poll_fn` calls are gone in my code in #5158.